Major improvement for syncset apply performance and simulated scale testing.

dgoodwin opened the pull request:
Adds a new developer annotation that can be applied to HiveConfig, hive.openshift.io/fake-applies: true, which will configure hive to fake out all Apply operations (SyncSets) to remote clusters. This is for simulated scale testing and allows us to use one spoke cluster and adopt it thousands of times. 

Commit still included for Greg's original work in this area in case we ever want to reference or use it, his solution uses a nodejs server which we tell to wait a specified ms before returning. In my testing a Sleep performs fairly similar, a little faster but I believe this is because we are double simulating requests in the nodejs approach, if we lookup real world data for the sleep array, then do an http request telling it to wait this time, it will be slower than the data we looked up. I believe Sleep is a reasonable simulation and much easier to commit in this fashion for re-use, however I wanted to keep the http request sim in the commit history.

Major fix in this PR is for performance, I found that we surprisingly still are doing OpenAPI schema builds on every single resource apply. We found this previously and it was the driver for moving to ClusterSync instead of SyncSetInstance controllers, however somewhere we missed this along the way. This operation took 3s when we first discovered it in the wild IIRC, and in my testing against one cluster it was taking 11s. Done 600 times, we can't complete a reconcile loop in less than an hour.

This fix moves the schema build up into the constructor for the resource apply Helper and then stores it as a field.

dgoodwin commented:
/assign @twiest 
/cc @csrwng 

Cesar could you verify the resource helper changes look ok to you?

csrwng commented on pkg/resource/apply.go:
```diff
@@ -41,21 +47,24 @@ const fieldTooLong metav1.CauseType = "FieldValueTooLong"
+	// fakeApply mode (used in scale testing only) doesn't run the apply. Instead it will Sleep for a period of time
+	// that matches our request time metrics in real hive environments to simulate i/o wait.
+	fakeAppliesEnvVar, _ := os.LookupEnv(constants.HiveSyncSetsFakeApplyEnvVar)
+	if fakeApplies, err := strconv.ParseBool(fakeAppliesEnvVar); fakeApplies && err == nil {
```
should the env var and and ParseBool be called when constructing the resource helper instead of every apply?

csrwng commented on pkg/resource/apply.go:
```diff
@@ -125,11 +123,29 @@ func (r *helper) CreateRuntimeObject(obj runtime.Object, scheme *runtime.Scheme)
 		return "", err
 	}
+	if info == nil {
+		r.logger.Debug("err getting info")
```
Should likely return an error here. Is this something you've observed? (where err == nil and info == nil?)

dgoodwin replied:
I honestly don't even remember adding this. I'll remove, probably something from when I was stuck trying to figure out what was wrong.

csrwng commented on pkg/resource/info.go:
```diff
@@ -21,11 +21,7 @@ type Info struct {
-		return nil, err
-	}
-	resourceInfo, err := r.getResourceInfo(factory, obj)
+	resourceInfo, err := r.getResourceInfo(r.factory, obj)
```
just `return r.getResourceInfo(r.factory, obj)`

csrwng commented on pkg/resource/helper.go:
```diff
@@ -44,59 +47,70 @@ type helper struct {
-		kubeconfig:     kubeconfig,
-		remote:         true,
+	// OpenAPISchema lookup is an extremely expensive operation, taking upwards of 3s often.
+	// We want to be very sure to only due this once per interaction with a cluster and as such,
```
s/due/do/

csrwng reviewed (commented):
Hey @dgoodwin, looks good. Some initial comments.
Just thinking about the fakeApply, I wonder if it gives you a realistic simulation of load given that regardless of the network communication, cpu cycles are normally spent on serializing/deserializing resources.

dgoodwin commented:
> Just thinking about the fakeApply, I wonder if it gives you a realistic simulation of load given that regardless of the network communication, cpu cycles are normally spent on serializing/deserializing resources.

Ideally we'd plug the fake apply in at a lower level and hit more serialization, but this appears to be as low as we can get without getting into modifying vendored code. I'm wondering if we're hitting most of the serialization though:

From prod cluster data, we can see that the average apply time is about 0.4s, average GET from clustersync controller is about 0.046, PATCH 0.047, POST 0.085. I'm not entirely sure what combo of those would get done for a normal apply, probably just a GET and a no-op? So network maybe 46ms out of 400ms total for an apply.

In our test cluster we're seeing applys take 70ms (although with beefier nodes, c5.4xlarge as opposed to m5.4xlarge), and simulating an avg around 46ms. This may indicate you are right. I'm going to try to get some data for pods running on m5.xlarge today and see how close we are to the apply time in prod. I'm not sure how we could get it more accurate, maybe try to replicate some of the expensive function calls up here in our code? Or maybe we just can't reliably commit this hack and need to do it by editing vendored code *when* we're doing scale testing.

abhinavdahiya commented:
qq: does fetching openAPI data once and re using affect list of manifests that install a CRD and the it's corresponding CR ? The clients usually use the openAPI data to figure out the api endpoint for a GroupVersionKind and old openAPI data will not have the new API created by the CRD ...

dgoodwin commented:
> qq: does fetching openAPI data once and re using affect list of manifests that install a CRD and the it's corresponding CR ? The clients usually use the openAPI data to figure out the api endpoint for a GroupVersionKind and old openAPI data will not have the new API created by the CRD ...

We've seen this fail even when we rebuilt the openapi schema every request as well, it seems sometimes in the past kube can be too slow to fully register a CRD and make it available, and if you try to use it on the next request it might fail saying no such resource.

This will likely be possible to surface but the clustersync should just requeue, which rebuilds the openapi schema and the next attempt should work. If you were doing a lot of CRDs + CRs, it could be a number of syncs for a successful reconcile. Definitely something to think about, but hopefully will be ok, and probably our only option, I don't think we can leave it the way it is. 

Perhaps a rebuild if we see we sync a CRD?

dgoodwin commented on pkg/resource/helper.go:
```diff
@@ -44,59 +47,70 @@ type helper struct {
-	r.getFactory = r.getKubeconfigFactory
-	return r
+	rh, _ := r.(*helper)
+	rh.metricsEnabled = true
```
Superb catch, this probably broke metrics.

dgoodwin commented:
PR is updated for feedback. I don't have a good idea yet on how to better simulate the serialization we're missing by bypassing some of the apply code. I think we can defer this for the time being and the current form as our baseline for now, with horizontal scaling pods we can mitigate any differences in time spent serializing if necessary.

dgoodwin commented:
/test integration

dgoodwin commented:
/hold
Legit failure, looks like some of the dropped kubeconfig factory code for the NewHelper(kubeconfig) variant was ruined in my refactor, I am trying to figure out how to restore.

csrwng commented on pkg/controller/utils/clientwrapper.go:
```diff
@@ -103,6 +110,24 @@ type ControllerMetricsTripper struct {
+	// "CancelRequest not implemented by *utils.ControllerMetricsTripper" was seen during simulated scale
+	// testing as the cluster was falling over. We will now implement this and track a metric for requests
+	// being cancelled as this could prove valuable in tracking when we have a performance issue.
+	remoteStr := "false"
```
nit: `remoteStr := strconv.FormatBool(cmt.Remote)`

csrwng commented on pkg/operator/hive/hive.go:
```diff
@@ -165,6 +171,14 @@ func (r *ReconcileHiveConfig) deployHive(hLog log.FieldLogger, h resource.Helper
+			Name:  hiveconstants.HiveSyncSetsFakeApplyEnvVar,
+			Value: "true",
+		})
+	}
```
should at least log an error if the value can't be parsed (or even better, just exit/crash)

dgoodwin replied:
This is an operator controller and we don't seem to have a precedent for crashing the pod if something is misconfigured for the hive controllers deployment, so maybe log and return an error here is best as that seems to be consistent.

csrwng reviewed (commented):
Current code looks good to me, a couple of nits, feel free to ignore.

dgoodwin commented on pkg/resource/patch.go:
```diff
@@ -27,11 +27,7 @@ func (r *helper) Patch(name types.NamespacedName, kind, apiVersion string, patch
 		Out:    &bytes.Buffer{},
 		ErrOut: &bytes.Buffer{},
 	}
-	factory, err := r.getFactory(name.Namespace)
```
This is where I got into trouble, I didn't notice that some of these getFactory commands were specifying a namespace, so my attempt to store the factory/dynClient/etc on the helper and re-use them was misguided. I'm reworking it to only cache the openapi schema.

dgoodwin commented:
PR is now updated, I had to rework the resource handler patch quite a bit, it's closer to what it was before now, with just a temporary factory used during resource helper constuction to get the openapi schema, and then that is all we cache on the helper. 

Also updated with your last suggestions @csrwng

csrwng commented:
/lgtm

dgoodwin commented:
/hold cancel
